taichi/program/async_engine.cpp

```diff
@@ -21,20 +21,53 @@ uint64 hash(IRNode *stmt) {
   std::string serialized;
   irpass::re_id(stmt);
   irpass::print(stmt, &serialized);
+  // TODO: separate kernel from IR template
+  serialized += stmt->get_kernel()->name;
   uint64 ret = 0;
   for (uint64 i = 0; i < serialized.size(); i++) {
     ret = ret * 100000007UL + (uint64)serialized[i];
   }
   return ret;
 }
 
-std::unique_ptr<OffloadedStmt> clone_offloaded_task(OffloadedStmt *from,
-                                                    Kernel *kernel) {
-  auto new_ir = irpass::analysis::clone(from, kernel);
-  return std::unique_ptr<OffloadedStmt>((OffloadedStmt *)(new_ir.release()));
+}  // namespace
+
+std::unique_ptr<IRNode> IRHandle::clone() const {
+  // TODO: remove get_kernel() here
+  return irpass::analysis::clone(const_cast<IRNode *>(ir_), ir_->get_kernel());
 }
 
-}  // namespace
+uint64 IRBank::get_hash(IRNode *ir) {
+  auto result_iterator = hash_bank_.find(ir);
+  if (result_iterator == hash_bank_.end()) {
+    auto result = hash(ir);
+    hash_bank_.insert(hash_bank_.end(), std::make_pair(ir, result));
+    return result;
+  }
+  return result_iterator->second;
+}
+
+void IRBank::set_hash(IRNode *ir, uint64 hash) {
+  hash_bank_[ir] = hash;
+}
+
+bool IRBank::insert(std::unique_ptr<IRNode> &&ir, uint64 hash) {
+  IRHandle handle(ir.get(), hash);
+  auto insert_place = ir_bank_.find(handle);
+  if (insert_place == ir_bank_.end()) {
+    ir_bank_.insert(ir_bank_.end(), std::make_pair(handle, std::move(ir)));
```

Comment:
nit: 
```suggestion
    ir_bank_.emplace(handle, std::move(ir));
```

---

Also, I wonder if `insert()` should always call `set_hash`?